Update coupled to latest UFS version

WalterKolczynski-NOAA opened the pull request:
Updates the coupled workflow to use the current version of UFS. This requires multiple updates to the configuration to handle new or changed settings.

Closes: #230, #324, #325, #333, #336, #337

rmontuoro committed:
Checkout and build ufs-weather-model app integrating
prognostic aerosols into the UFS, replacing original
FV3-GOCART system.

A forked repository is temporarily used to download
the updated ufs-weather-model app.

rmontuoro committed:
Use proper modulefiles for updated ufs-weather-model
app with prognostic aerosols.

rmontuoro committed:
Update parsing function for FV3 namelist to work
with latest FV3 revision and include settings for
FV3_GFS_v16 and aerosol scavenging factors.

rmontuoro committed:
Add aerosol scavenging factors (fscav_aero) to FV3 settings.
This item is implemented as a YAML list of strings, which
allows to set scavenging factors for each tracer.

rmontuoro committed:
Properly set NST_SPINUP flag in FV3 namelist.

rmontuoro committed:
Restore `cplchem` as the flag for coupled aerosols,
removing all references to `cplgocart`.

rmontuoro committed:
- Set NEMS identifier for coupled aerosol component to 'gocart'.

- Copy NUOPC field dictionary to run directory by default.

rmontuoro committed:
Update test cases and input files for coupled aerosols runs.

rmontuoro committed:
Add scavenging factors to aerosol test cases.

rmontuoro committed:
Use updated modulefile names only when building the
ufs-weather-model with prognostic aerosols. This
option is recognized by means of a hidden file
created in the model's repository when running
checkout.sh with option -a.

rmontuoro committed:
Load updated modulefiles for coupled aerosols on Hera/Intel.

A new shell function (get_platform) is introduced to increase
modularity. This fuction auto-detects the local platform and
returns a string identifier. As a result, load_fv3gfs_modules.sh
is refactored.

rmontuoro committed:
Remove GSD-specific functions in forecast postdet.

rmontuoro committed:
Remove no longer needed parsing script used to
generate the FV3 namelist file for FV3-GOCART.

rmontuoro committed:
Return proper name for Hera platform and load correct modules.

rmontuoro committed:
Load aerosol-specific modulefiles only for the forecast task.

rmontuoro committed:
Merge branch 'WalterKolczynski-NOAA:feature/aerosol_restart' into feature/update-aerosols

rmontuoro committed:
Replace legacy 'cplgocart' flag with 'cplchem'.

rmontuoro committed:
Update tracer list.

rmontuoro committed:
Merge branch 'WalterKolczynski-NOAA:feature/aerosol_restart' into feature/update-aerosols

rmontuoro committed:
Restore original path to modulefiles in module_base.hera.

rmontuoro committed:
Merge 'WalterKolczynski-NOAA:feature/aerosol_restart' into feature/update-aerosols

rmontuoro committed:
Update aerosol_firex_forecast test case.

WalterKolczynski-NOAA committed:
Update UFS model hash

Updates the UFS model to the current tip of develop.

Refs: #324

WalterKolczynski-NOAA committed:
Update MOM_input with landmask change option

Adds the ALLOW_LANDMASK_CHANGES option to the MOM6 input templates
and allows the setting via MOM6_ALLOW_LANDMASK_CHANGES (default
False).

Refs: #324

WalterKolczynski-NOAA committed:
Update CICE inputs for new model version

The newest version of CICE replaces istep0 with the normal month and day,
so the namelist is updated to reflect that.

ktherm and tfrz_option are updated to allow alternate settings, following
similar changes to the UFS test harness. The default values are the
previous settings.

Settings for tr_FY, sw_redist, f_mlt_onset, and f_frx_onset are updated
to new values in the UFS test harness.

Additionally, many namelist options have been removed following the same
deletions in the UFS test harness.

Refs: #324

WalterKolczynski-NOAA committed:
Update UFS build script

Some changes to the UFS require updates to the build script.

The modulefile organization has been updated to be
modulefiles/ufs_${target} rather than modulefiles/${target}/fv3, so
the module paths in the build script are updated correspondingly.

A cmake switch DAPP is now required in order to build UFS. Non-
aerosol coupled is using S2SW for the time being. This replaces
setting DS2S and DWW3 to YES.

The old v15 suite is removed from the CCPP suite list.

Refs: #324

WalterKolczynski-NOAA committed:
Update source of fd_nems.yaml

The location of the fd_nems.yaml file changed in the UFS repository,
so updated nems_configure to copy from the new location.

Refs: #324

WalterKolczynski-NOAA committed:
Update templates to set ocean mesh file

The latest version of UFS adds a setting to define the ocean mesh file,
so the nems.configure templates are updated to define them. They are
identical to the ice mesh file, so the MESHICE variable is renamed to
MESH_OCN_ICE and used for both. Additionally, the use_mommesh variable
is added and defaults to true.

Refs: #324

WalterKolczynski-NOAA committed:
Update Ww3 prepost build

The WW3 pre and post components now need NetCDF to build, so NetCDF is
added back to the modulelist, and previously removed code in the build
script is restored.

Refs: #324

JessicaMeixner-NOAA committed:
update wave ww3_multi.inp file Issue #325

WalterKolczynski-NOAA committed:
Update meditaor name

In the latest UFS, the mediator name was changed from nems to cmeps, so
the nems.configure is updated to match.

Refs: #324

WalterKolczynski-NOAA committed:
Update wave IC

Update wave IC to new set compatible with current UFS version.

Refs: #324

WalterKolczynski-NOAA committed:
Update modules to match UFS

Update module versions to match those in the UFS common file.

Refs: #324

WalterKolczynski-NOAA committed:
Update ice configuration to match UFS

There were still a number of settings in the ice input file that
did not match the setting in UFS (and one setting that was misplaced).

Refs: #324

WalterKolczynski-NOAA committed:
Merge branch 'feature/coupled-crow' into feature/model_update

WalterKolczynski-NOAA commented:
Currently on hold until ufs-community/ufs-weather-model/pull/664 is merged to fix a CICE issue, at which point the checkout hash will need to be updated.

JessicaMeixner-NOAA commented on sorc/build_ufs_coupled.sh:
```diff
@@ -13,15 +13,17 @@ if [ $target = hera ]; then target=hera.intel ; fi
+module load ufs_${target}
 cd ufs_coupled.fd/
 if [[ -d build ]]; then rm -Rf build; fi
 if [[ -d GOCART ]]; then
```
based on the https://github.com/NOAA-EMC/WW3/wiki/Commit-Queue the if statement for if the the GOCART directory exists is not going to work by the time the update for the ice history is merged in.

WalterKolczynski-NOAA replied:
I'll change this back to using the -a switch. Unless after ufs-community/ufs-weather-model/pull/591 there will be a DAPP option that incorporates everything? That would be ideal.

JessicaMeixner-NOAA replied:
Looks to me like: ATMAERO was added only.  Seems like a good APP/feature to maybe request for the future?

JessicaMeixner-NOAA commented on ush/parsing_namelists_CICE.sh:
```diff
@@ -388,11 +209,11 @@ cat > ice_in <<eof
    f_hs            = 'mdhxx' 
    f_Tsfc          = 'mdhxx' 
-   f_sice          = 'x' 
+   f_sice          = 'mdhxx' 
```
@DeniseWorthen @LydiaStefanova-NOAA Were these fields we had specifically turned off in the workflow?  Should we keep these off in the workflow or turn them back on?

LydiaStefanova-NOAA replied:
I can't answer this - I don't know what these settings do, nor what they should be. Noticed in the configuration spreadsheet that a number of them were changed from 'mdhxx' to 'x' between P3.1 and P4, but don't know why.

JessicaMeixner-NOAA replied:
@LydiaStefanova-NOAA this is for which fields we want in the ice history output files.  'x' means it's not included and then 'mdhxx' means we will have output for this.   These were turned off as you mentioned between P3.1 and P4 for the workflow, but have remained in the regression tests as its helpful for debugging.

WalterKolczynski-NOAA replied:
There were a number of these types of changes. At first I just changed the ones that seem to have been updated since the last commit, then I noticed other settings also didn't match an naively updated them. We can revert 2dcf8f2b, which is the one where I made that second batch of changes (and then re-do the one mispaste of the new sw_redist setting that was also fixed there). Take a look at the changes from just that commit and let me know.

JessicaMeixner-NOAA replied:
@WalterKolczynski-NOAA it's a lot easier when these files just match up exactly.  I do want to say this saved us a fair amount of space on benchmark runs though.  @LydiaStefanova-NOAA I guess the question is, do we want the extra output for ice files in the full 168 runs, or do we want to keep the list that was trimmed down in P4 for ice?  Should we check with @DeniseWorthen Sulagna and @rgrumbine and others looking at ice output to see if we can continue without the extra ice output or if it'd be good to add these back -- if for nothing else to have more consistency with the rt system.

LydiaStefanova-NOAA replied:
I went through the ice.nml section to list what's different from P6 - see spreadsheet [here](https://docs.google.com/spreadsheets/d/1CpJ7U33wIA-uWlqFewgBNbQ8AtAw0727d9c49QMiHnM/edit?usp=sharing). Red=newly turned on; Blue=newly turned off.

I went by [https://cesmcice.readthedocs.io/en/latest/users_guide/ice_history.html](url) to tie the parameters to descriptions. A couple of those were not on the list (f_dsnow and f_fbot). 

Of the newly turned off fields (8), we definitely want f_mlt_onset and f_frz_onset back on. Maybe f_divu and f_shear too. 

For the newly turned on fields (20) - most of them relate to radiation/albedo, salt, snow, ocean currents, atm winds. They may be useful to someone for debugging/sanity checking - I will leave it to others to weigh on that.

WalterKolczynski-NOAA replied:
Per later comment from Jessica, all those that were turned off have been turned back on.

rmontuoro committed:
Merge 'feature/coupled-crow' into feature/update-aerosols

rmontuoro committed:
Add ammonia and nitrates to the list of cycled GOCART tracers.

rmontuoro committed:
Update checkout and build script to use the authoritative
GitHub repository for the UFS weather model only. This repo
now contains UFS-Aerosols, the aerosol component based on
NASA's GOCART.

Since we are using a single repository for both S2S and
coupled aerosols, the -a option was removed from checkout.sh
and added to build_all.sh. This option is passed to
build_ufs_coupled.sh and partial_build.sh, which have
been streamlined and now accept both short and long options:

build_ufs_coupled.sh:
  -a, --aerosols
  -c, --coupled

partial_build.sh:
  -a, --aerosols
  -c, --coupled
  -h, --help
  -v, --verbose

KateFriedman-NOAA committed:
Add spaces to indent FV3 build section for better readability

Refs: #230

KateFriedman-NOAA committed:
Add FV3_GFS_v16_couplednsst CCPP suite option to UFS coupled build and FV3 namelist parser

Refs: #230

KateFriedman-NOAA committed:
Replace CROW yaml nst_spinup_logic in nsst.yaml with NST_SPINUP variable

- removed CROW FirstTrue check when building config.nsst from nsst.yaml
- replace nst_spinup_logic with NST_SPINUP variable that is parsed from variable dictionary
- nst_spinup_logic not needed currently

Refs: #230

KateFriedman-NOAA committed:
Add NSST settings to the coupled_free_forecast_wave.yaml case file

- set NST_MODEL=2 (turns it on)
- add additional NSST namelist variables: NST_SPINUP, NST_RESV, ZSEA1, ZSEA2
- set NST_SPINUP=1 (NSST will spinup)
- change CCPP_SUITE from FV3_GFS_v16_coupled to FV3_GFS_v16_couplednsst

Refs: #230

WalterKolczynski-NOAA committed:
Update nems.configure to inclue eps_imesh

Adds the eps_imesh setting to the nems.configure files when CICE is
on.

Refs: #324, #333

WalterKolczynski-NOAA committed:
Merge branch 'feature/model_update' into feature/update-aerosols

WalterKolczynski-NOAA committed:
Change forecast module back to module_base

The modulefile in UFS doesn't load some modules needed for the forecast
script at runtime, notably produtil, which provides err_chk. The module
for the forecast job is changed back to module_base for the time being.

Refs: #336

JessicaMeixner-NOAA commented on ush/parsing_namelists_CICE.sh:
```diff
@@ -435,51 +256,51 @@ cat > ice_in <<eof
    f_strength      = 'x'
-   f_divu          = 'mdhxx'
-   f_shear         = 'mdhxx'
+   f_divu          = 'x'
```
let's revert any that have gone from "on" to "off" (mdhxx->x), but anything that was turned on (x->mdhxx) as on.  

So the variables to revert are: f_divu, f_shear, f_divdt* (lines 296-299), f_mld_onset, f_frz_onset

WalterKolczynski-NOAA replied:
Done

WalterKolczynski-NOAA committed:
Update UFS to 20210720 develop

Updates the UFS version to the current develop.

The UFS build script in g-w is changed as the bulid script in the UFS
repo no longer automatically adds CCPP_FLAGS to CMAKE_FLAGS, so now the
g-w version does it instead before calling the UFS script. While this
change was being made, the option parsing was updated to use getopts to
match other scripts.

The FV3 namelist has ncld removed as that option is no longer recognized.

Refs: #324

WalterKolczynski-NOAA committed:
Revert some of the CICE in changes

Some of the fields that had previously been turned off from output
in CICE are turned back on.

Refs: #324, #333

WalterKolczynski-NOAA commented:
coupled_free_forecast_wave, aerosol_firex_forecast, and aerosol_free_forecast all report as succeeded on Hera.

JessicaMeixner-NOAA commented on sorc/build_all.sh:
```diff
@@ -9,14 +9,25 @@ set +x
+while getopts "ac" option;
 do
  case $option in
+  a)
```
Does the readme need to be updated for this?

WalterKolczynski-NOAA replied:
Yes. I'll get on that.

WalterKolczynski-NOAA replied:
Updated

WalterKolczynski-NOAA committed:
Update README with aerosol build option

README is updated with instructions to build UFS for aerosols.

Refs: #324, #336

JessicaMeixner-NOAA reviewed (approved):
For the coupled side of things, this all looks fine.  Approving based on the fact that @WalterKolczynski-NOAA had a successful test on hera.

rmontuoro reviewed (approved):
Looks good.
